botocore/credentials.py

```diff
@@ -835,161 +1016,295 @@ def __init__(self, load_config, client_creator, cache, profile_name,
         # to load credentials (as opposed to when the object is
         # instantiated).
         self._loaded_config = {}
+        self._credential_sourcer = credential_sourcer
+        self._visited_profiles = [self._profile_name]
 
     def load(self):
         self._loaded_config = self._load_config()
         if self._has_assume_role_config_vars():
-            return self._load_creds_via_assume_role()
+            return self._load_creds_via_assume_role(self._profile_name)
 
     def _has_assume_role_config_vars(self):
         profiles = self._loaded_config.get('profiles', {})
         return self.ROLE_CONFIG_VAR in profiles.get(self._profile_name, {})
 
-    def _load_creds_via_assume_role(self):
-        # We can get creds in one of two ways:
-        # * It can either be cached on disk from an pre-existing session
-        # * Cache doesn't have the creds (or is expired) so we need to make
-        #   an assume role call to get temporary creds, which we then cache
-        #   for subsequent requests.
-        creds = self._load_creds_from_cache()
-        if creds is not None:
-            logger.debug("Credentials for role retrieved from cache.")
-            return creds
-        else:
-            # We get the Credential used by botocore as well
-            # as the original parsed response from the server.
-            creds, response = self._retrieve_temp_credentials()
-            cache_key = self._create_cache_key()
-            self._write_cached_credentials(response, cache_key)
-            return creds
-
-    def _load_creds_from_cache(self):
-        cache_key = self._create_cache_key()
-        try:
-            from_cache = self.cache[cache_key]
-            if self._is_expired(from_cache):
-                # Don't need to delete the cache entry,
-                # when we refresh via AssumeRole, we'll
-                # update the cache with the new entry.
-                logger.debug(
-                    "Credentials were found in cache, but they are expired.")
-                return None
-            else:
-                return self._create_creds_from_response(from_cache)
-        except KeyError:
-            return None
-
-    def _is_expired(self, credentials):
-        end_time = _parse_if_needed(credentials['Credentials']['Expiration'])
-        now = datetime.datetime.now(tzlocal())
-        seconds = total_seconds(end_time - now)
-        return seconds < self.EXPIRY_WINDOW_SECONDS
+    def _load_creds_via_assume_role(self, profile_name):
+        role_config = self._get_role_config(profile_name)
+        source_credentials = self._resolve_source_credentials(
```

Comment:
This resolution is now happening before the cache checks.  This results in an unnecessary requirement that the entire chain of assumed profiles is not expired.

However, these intermediate results aren't actually used if a subsequent role *is* cached.  For example, consider the role chain `A->B->C`.  Let's say C  and B are expired in the cache, but A's entry in `~/.aws/cli/cache` is up to date.  This means the sequence of calls you'll make is:

* `AssumeRole(C)`
* `AssumeRole(B)`
* Used cache credentials from A

It seems like the ideal behavior would be if we're trying to assume the role in profile `A`, as long as it's cache entry is up to date we should just use it.

You can test this behavior by just deleting the cache file for `B` and `C` in `~/.aws/cli/cache`

Reply:
Looks like this crept in when moving from assume role credentials to the fetcher since I didn't have a test for it. The problem is that `RefreshableCredentials` needs to have initial credentials.